### PR TITLE
Rework internal TokenStream representation

### PR DESCRIPTION
**Pull request by maciektr**
commit-id:ab3e2fd1

---

**Stack**:
- #1734
- #1722
- #1704
- #1700
- #1699 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

**Comment by maciektr** [comment]
This is just #1639 but moved to a stack

File: plugins/cairo-lang-macro/Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,7 @@ repository.workspace = true
 
 [dependencies]
 cairo-lang-macro-attributes = "0.1"
-cairo-lang-macro-stable = "1"
+cairo-lang-macro-stable = { path = "../cairo-lang-macro-stable" }
```
Comment (wawel37) [inline plugins/cairo-lang-macro/Cargo.toml:17]: Just a note to change it during the release

File: scarb/Cargo.toml [hunk 2]
```diff
@@ -23,7 +23,7 @@ cairo-lang-filesystem.workspace = true
 cairo-lang-formatter.workspace = true
 cairo-lang-lowering.workspace = true
 cairo-lang-macro = { path = "../plugins/cairo-lang-macro" }
-cairo-lang-macro-stable = "1"
+cairo-lang-macro-stable = { path = "../plugins/cairo-lang-macro-stable" }
```
Comment (wawel37) [inline scarb/Cargo.toml:26]: Same as above

**Review by wawel37 — APPROVED** [review]

File: plugins/cairo-lang-macro/src/types/token.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+pub struct TextSpan {
+    pub start: usize,
+    pub end: usize,
+}
+
+/// A single Cairo token.
+///
+/// The most atomic item, of Cairo code representation, when passed between macro and host.
```
Comment (wawel37) [inline plugins/cairo-lang-macro/src/types/token.rs:36]:
```suggestion
/// The most atomic item of Cairo code representation.
```

File: plugins/cairo-lang-macro/src/types/token.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+pub struct Token {
+    pub content: String,
+    pub span: TextSpan,
+}
+
+/// Metadata of [`TokenStream`].
+///
+/// This struct can be used to describe the origin of the [`TokenStream`].
```
Comment (wawel37) [inline plugins/cairo-lang-macro/src/types/token.rs:46]:
```suggestion
/// This struct describes the origin of the [`TokenStream`].
```

**Review by wawel37 — APPROVED** [review]

**Pull request merged**